flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java

```diff
@@ -308,4 +470,88 @@ private static void writeFileAtomically(
 
 		assertTrue(stagingFile.renameTo(file));
 	}
+
+	private enum FailoverType {
+		NONE,
+		TM,
+		JM
+	}
+
+	private static class HaServices extends TestingEmbeddedHaServices {
+		private final Supplier<CompletedCheckpointStore> completedCheckpointStoreFactory;
+		private final CheckpointIDCounter checkpointIDCounter;
+
+		private HaServices(
+				Executor executor,
+				Supplier<CompletedCheckpointStore> completedCheckpointStoreFactory,
+				CheckpointIDCounter checkpointIDCounter) {
+			super(executor);
+			this.completedCheckpointStoreFactory = completedCheckpointStoreFactory;
+			this.checkpointIDCounter = checkpointIDCounter;
+		}
+
+		@Override
+		public CheckpointRecoveryFactory getCheckpointRecoveryFactory() {
+			return new CheckpointRecoveryFactoryWithSettableStore(
+				completedCheckpointStoreFactory,
+				checkpointIDCounter);
+		}
+	}
+
+	private static class CheckpointRecoveryFactoryWithSettableStore implements CheckpointRecoveryFactory {
+		private final Supplier<CompletedCheckpointStore> completedCheckpointStoreFactory;
+		private final CheckpointIDCounter checkpointIDCounter;
+
+		private CheckpointRecoveryFactoryWithSettableStore(
+				Supplier<CompletedCheckpointStore> completedCheckpointStoreFactory,
+				CheckpointIDCounter checkpointIDCounter) {
+			this.completedCheckpointStoreFactory = completedCheckpointStoreFactory;
+			this.checkpointIDCounter = checkpointIDCounter;
+		}
+
+		@Override
+		public CompletedCheckpointStore createCheckpointStore(
+				JobID jobId,
+				int maxNumberOfCheckpointsToRetain,
+				ClassLoader userClassLoader) {
+			return completedCheckpointStoreFactory.get();
+		}
+
+		@Override
+		public CheckpointIDCounter createCheckpointIDCounter(JobID jobId) {
+			return checkpointIDCounter;
+		}
+	}
+
+	private enum RecordCounterToFail {
```

Comment:
Can we change this to a regular static class? I though the consensus was to not use the enum pattern any more, because there is not really a problem that it fixes and it causes confusion for many developers.